crates/symbolicator-proguard/src/java.rs

```diff
@@ -0,0 +1,229 @@
+use std::collections::HashMap;
+use std::sync::OnceLock;
+
+fn java_base_types() -> &'static HashMap<String, String> {
+    static JAVA_BASE_TYPES: OnceLock<HashMap<String, String>> = OnceLock::new();
+    JAVA_BASE_TYPES.get_or_init(|| {
+        // https://docs.oracle.com/javase/8/docs/jdk/api/jpda/jdi/com/sun/jdi/doc-files/signature.html
+        HashMap::from([
+            ("Z".to_string(), "boolean".to_string()),
+            ("B".to_string(), "byte".to_string()),
+            ("C".to_string(), "char".to_string()),
+            ("S".to_string(), "short".to_string()),
+            ("I".to_string(), "int".to_string()),
+            ("J".to_string(), "long".to_string()),
+            ("F".to_string(), "float".to_string()),
+            ("D".to_string(), "double".to_string()),
+            ("V".to_string(), "void".to_string()),
+        ])
+    })
+}
+
+fn byte_code_type_to_java_type(
+    byte_code_type: String,
+    mappers: &[&proguard::ProguardMapper],
+) -> String {
+    if byte_code_type.is_empty() {
+        return "".to_string();
+    }
+    let mut chrs = byte_code_type.chars();
+    let token = chrs.next().unwrap();
+    if let Some(ty) = java_base_types().get(&token.to_string()) {
+        return ty.clone();
+    } else if token == 'L' {
+        // invalid signature
+        let l = chrs.clone().last();
+        if l.is_none() || l.unwrap() != ';' {
+            return "".to_string();
+        }
+        chrs.next_back(); // remove final `;`
+        let obfuscated = chrs.collect::<String>().replace('/', ".");
+        // if mappers.len() > 0 {}
+        for mapper in mappers {
+            if let Some(mapped) = mapper.remap_class(&obfuscated) {
+                return mapped.to_string();
+            }
+        }
+        return obfuscated;
+    } else if token == '[' {
+        return format!(
+            "{}[]",
+            byte_code_type_to_java_type(chrs.collect::<String>(), mappers)
+        );
+    }
+    byte_code_type
+}
+
+// parse_obfuscated_bytecode_signature will parse an obfuscated signatures into parameter
+// and return types that can be then deobfuscated
+fn parse_obfuscated_bytecode_signature(signature: &str) -> Option<(Vec<String>, String)> {
+    let mut chrs = signature.chars();
+
+    let token = chrs.next();
+    if token.unwrap_or_default() != '(' {
+        return None;
+    }
+
+    let sig = chrs.collect::<String>();
+    let split_sign = sig.rsplitn(2, ')').collect::<Vec<&str>>();
+    if split_sign.len() != 2 {
+        return None;
+    }
```

Comment:
Isn’t there also an `rsplit` that returns a `Option<(&str,&str)>`?

Other than that, you might want to use `strip_prefix` to remove that leading `(`.

Reply:
@Swatinem `rsplitn` is restricted to returning at most **n** items. `rsplit` does not ensure that.

Also, here we want to check specifically if the first token is not a `(` as it would be an invalid signature. And with `token.unwrap_or_default() != '('` we're both checking and consuming so no need for removing that later.